### DISCUSSION TITLE
이벤트 핸들러에 타입을 주는 3가지 방법

### DISCUSSION THREAD

**Discussion by bassyu** [woowacourse-teams/2023-pium#59] [뜨거운 토론]
이벤트 핸들러를 만들 때 아래처럼 다양하게 타입을 줄 수 있습니다. 
3가지 방법중 하나로 통일해봐도 좋을 것 같아서 글 올립니다!

방법은 아래 3가지가 있습니다.
3개 모두 동일한 이벤트 객체 타입을 가집니다.

## 1. 함수 파라미터에 `이벤트 객체` 정의

```tsx
interface Props {
  handle: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const handle = (event: React.ChangeEvent<HTMLInputElement>) => {};
```

## 2. 함수에 `이벤트 핸들러` 정의

```tsx
interface Props {
  handle: React.ChangeEventHandler<HTMLInputElement>;
}

const handle: React.ChangeEventHandler<HTMLInputElement> = (event) => {};
```

## 3. 함수에 `태그의 이벤트 핸들러 속성` 정의

```tsx
interface Props {
  handle: React.ComponentProps<'input'>['onChange'];
}

const handle: React.ComponentProps<'input'>['onChange'] = (event) => {};
```

어떤 이벤트인지 생각할 필요 없이 태그와 속성만 알면 됩니다. 가독성이 좋아보여요!
대신 뾰족한 만큼 비슷한 이벤트를 컴포넌트끼리 공유할 때는 따로 작성해야하는 불편함이 있겠네요.

```tsx
const handle: React.ComponentProps<'input' | 'button' | 'div'>['onClick'] = (event) => {};
```

## 4. 타입 통일하지 않기

각자 편한걸 쓰자는 의견도 있을 것 같습니다.

**Comment by WaiNaat** [comment]
4가 괜찮은 것 같아요

어차피 [이름이 잘 지어진 이벤트 핸들러](https://blog.shiren.dev/2020-07-27-1/)라면 뾰족하지 않아도 헷갈릴 이유가 없어 보이는데 굳이 통일할 이유를 느끼지 못했습니다
